### PR TITLE
Handwritten adjoints and removal of Zygote

### PR DESCRIPTION
**Pull request by michel2323**
This PR introduces handwritten adjoints for all the constraints. The line flow constraint adjoints exhibit the same issues than the Zygote ones: They disagree with ForwardDiff and FD about `1e-4`. This issue has to be kept open. It is either a bug or some unknown numerical issue.

### CODE REVIEW

File: src/polar/kernels.jl [hunk 1]
```diff
@@ -244,6 +307,52 @@ function transfer!(polar::PolarForm, buffer::PolarNetworkState, u)
     wait(ev)
 end
 
+KA.@kernel function adj_transfer_kernel!(
+    adj_vmag, adj_vang, adj_pinj, adj_qinj, adj_u, pv, pq, ref, adj_pload, adj_qload
```
Comment (frapac) [inline src/polar/kernels.jl:311]: I guess here `adj_pload` is the adjoint of the power generation `pg` (`adj_pg`), as the load is constant?

File: src/polar/kernels.jl [hunk 2]
```diff
@@ -130,14 +133,14 @@ KA.@kernel function residual_adj_kernel!(F, adj_F, v_m, adj_v_m, v_a, adj_v_a,
     if i > npv
         F[i + npq] -= qinj[fr]
     end
-    @inbounds for c in ybus_re_colptr[fr]:ybus_re_colptr[fr+1]-1
+    @inbounds for c in colptr[fr]:colptr[fr+1]-1
```
Comment (frapac) [inline src/polar/kernels.jl:136]: Nice cleansing!

File: src/polar/kernels.jl [hunk 3]
```diff
@@ -181,22 +192,74 @@ KA.@kernel function residual_adj_kernel!(F, adj_F, v_m, adj_v_m, v_a, adj_v_a,
...
+
+function adj_residual_polar!(F, adj_F, vm, adj_vm, va, adj_va,
                          ybus_re, ybus_im,
                          pinj, adj_pinj, qinj, pv, pq, nbus)
     npv = length(pv)
     npq = length(pq)
+    nvbus = length(vm)
+    nnz = length(ybus_re.nzval)
+    T = eltype(vm)
+    edge_vm_from = zeros(T, nnz)
```
Comment (frapac) [inline src/polar/kernels.jl:226]: Do you know if we could reuse these variables to compute the adjoint of the line flow constraints?

**Review by frapac — COMMENTED** [review]
Just some preliminary comments on this PR. I need to write down the math to better understand what we are doing :grimacing:

File: src/polar/kernels.jl (hunk 3 above)
Comment (michel2323) [reply src/polar/kernels.jl]: Yes we could. I guess this belongs into the cache together with the adjoints.

**Comment by michel2323** [comment]
> Just some preliminary comments on this PR. I need to write down the math to better understand what we are doing 😬

I take shortcuts in the code; like for example when to use `+=` etc. This is not the best example to learn manual differentiation.

File: src/polar/kernels.jl (hunk 1 above)
Comment (michel2323) [reply src/polar/kernels.jl]: I am not sure. In that case we could remove that argument. I am not yet sure how to handle the `adjoint!` signatures in general. What we also have to clarify is weither the adjoint routines return a primal value. I would not recommend that.

File: .github/workflows/action.yml [hunk 4]
```diff
@@ -20,6 +20,8 @@ jobs:
       - uses: julia-actions/julia-runtest@latest
 
   test-moonshot:
+    env:
+      CUDA_VISIBLE_DEVICES: 1
```
Comment (frapac) [inline .github/workflows/action.yml:24]: Why setting this value? (I am curious :smiley_cat: )

File: src/Polar/Constraints/line_flow.jl [hunk 5]
```diff
@@ -24,46 +24,25 @@ function flow_constraints(polar::PolarForm, cons, vm, va, pbus, qbus)
 end
 
 function flow_constraints_grad!(polar::PolarForm, cons_grad, buffer, weights)
-    T = typeof(buffer.vmag)
     if isa(buffer.vmag, Array)
-        kernel! = accumulate_view!(KA.CPU())
+        kernel! = adj_branch_flow_kernel!(KA.CPU())
```
Comment (frapac) [inline src/Polar/Constraints/line_flow.jl:28]: When `polar` is available in the scope, I would suggest defining the kernel directly as
```julia
kernel! = adj_branch_flow_kernel!(polar.device)
```

File: src/Polar/kernels.jl [hunk 6]
```diff
@@ -1,4 +1,5 @@
 import KernelAbstractions: @index
+using ForwardDiff
```
Comment (frapac) [inline src/Polar/kernels.jl:2]: I don't think we need to import `ForwardDiff` here, as it is already loaded in `src/ExaPF.jl`

File: src/Polar/kernels.jl [hunk 7]
```diff
@@ -404,51 +747,70 @@ KA.@kernel function branch_flow_kernel!(
...
-end
+    # branch apparent power limits - from bus
+    yff_abs = yff_re[ℓ]^2 + yff_im[ℓ]^2
+    yft_abs = yft_re[ℓ]^2 + yft_im[ℓ]^2
+    yre_fr =   yff_re[ℓ] * yft_re[ℓ] + yff_im[ℓ] * yft_im[ℓ]
+    yim_fr = - yff_re[ℓ] * yft_im[ℓ] + yff_im[ℓ] * yft_re[ℓ]
 
-"""
-    accumulate_view(x, vx, indices)
+    # not needed in the reverse run
```
Comment (frapac) [inline src/Polar/kernels.jl:770]: Do we need this dead comment?

File: src/Polar/kernels.jl [hunk 8]
```diff
@@ -404,51 +747,70 @@ KA.@kernel function branch_flow_kernel!(
...
-KA.@kernel function accumulate_view!(x, vx, indices)
-    # This is parallelizable
-    id = @index(Global, Linear)
-    for (j, idx) in enumerate(indices)
-        if id == idx
-            x[id] += vx[j]
-        end
-    end
+    # not needed in the reverse run
+    # to_flow = vmag[to_bus]^2 * (
```
Comment (frapac) [inline src/Polar/kernels.jl:784]: and this one?

File: src/Polar/kernels.jl [hunk 9]
```diff
@@ -404,51 +747,70 @@ KA.@kernel function branch_flow_kernel!(
...
+    #     ytf_abs * vmag[fr_bus]^2 + ytt_abs * vmag[to_bus]^2 +
+    #     2 * vmag[fr_bus] * vmag[to_bus] * (yre_to * cosθ - yim_to * sinθ)
+    # )
+    # slines[ℓ + nlines] = to_flow
+
+    adj_to_flow = adj_slines[ℓ + nlines]
+    adj_vmag[to_bus] += (2 * vmag[to_bus] * ytf_abs * vmag[fr_bus]^2
```
Comment (frapac) [inline src/Polar/kernels.jl:791]: I would replace `2` by `2.0` (same `4` and `6`)

File: src/Polar/kernels.jl [hunk 10]
```diff
@@ -404,51 +747,70 @@ KA.@kernel function branch_flow_kernel!(
...
+    adj_vmag[to_bus] += (2 * yft_abs * vmag[fr_bus]^2 * vmag[to_bus]
+                       + 2 * vmag[fr_bus]^3 * (yre_fr * cosθ - yim_fr * sinθ)
+                        ) * adj_from_flow
+    adj_cosθ += 2.0 * vmag[to_bus] * vmag[fr_bus]^3 *   yre_fr  * adj_from_flow
+    adj_sinθ += 2.0 * vmag[to_bus] * vmag[fr_bus]^3 * (-yim_fr) * adj_from_flow
+
+    adj_Δθ =   cosθ * adj_sinθ
+    adj_Δθ -=  sinθ * adj_cosθ
+    adj_vang[fr_bus] += adj_Δθ
+    adj_vang[to_bus] -= adj_Δθ
```
Comment (frapac) [inline src/Polar/kernels.jl:815]: Nice job! This adjoint does not look trivial

**Review by frapac — COMMENTED** [review]
Ok to merge once minor comments have been addressed

File: src/Polar/kernels.jl (hunk 10 above)
Comment (michel2323) [reply src/Polar/kernels.jl:815]: That's why it's off!

File: src/Polar/kernels.jl (hunk 9 above)
Comment (michel2323) [reply src/Polar/kernels.jl:791]: You are right. I tried that actually yesterday to make sure that's not the cause of the numerical issues.

File: src/Polar/kernels.jl (hunk 7 above)
Comment (michel2323) [reply src/Polar/kernels.jl:770]: I prefer to have the forward code as a comment in the adjoint function. That's just a preference, we can remove it if you don't like it. It's easier for readability of the adjoint code if someone looks at it.

File: src/Polar/kernels.jl (hunk 6 above)
Comment (michel2323) [reply src/Polar/kernels.jl:2]: ok!

File: src/Polar/Constraints/line_flow.jl (hunk 5 above)
Comment (michel2323) [reply src/Polar/Constraints/line_flow.jl:28]: oh nice! indeed!

File: src/Polar/kernels.jl (hunk 7 above)
Comment (frapac) [reply src/Polar/kernels.jl:770]: Ok to keep the comment, then!

File: .github/workflows/action.yml (hunk 4 above)
Comment (michel2323) [reply .github/workflows/action.yml:24]: This makes the tests on moonshot run on `device=1` to avoid a clash with most users and #110

**Pull request merged**